src/splitter.rs

```diff
@@ -15,8 +17,84 @@ impl Splitter {
             .collect()
     }
 
-    pub fn phrases(&self, _text: &str) -> Vec<Vec<u16>> {
-        todo!("Implement phrases extraction")
+    pub fn phrases(&self, text: &str) -> Vec<Vec<u16>> {
+        let string_phrases = self.phrases_as_strings(text);
+        let vocabulary = self.vocabulary(&self.clean_text(text));
+        
+        string_phrases.into_iter()
+            .map(|phrase| self.phrase_to_indices(&phrase, &vocabulary))
+            .collect()
+    }
+
+    pub fn phrases_as_strings(&self, text: &str) -> Vec<Vec<String>> {
+        self.split_into_sentences(text)
+            .into_iter()
+            .map(|sentence| self.clean_sentence(&sentence))
+            .filter(|words| words.len() >= 2)
+            .flat_map(|words| self.generate_substrings(&words))
+            .collect::<BTreeSet<_>>()
+            .into_iter()
+            .collect()
+    }
+    
+    fn split_into_sentences(&self, text: &str) -> Vec<String> {
+        text.split("\n\n")
+            .flat_map(|paragraph| {
+                paragraph.split(|c| matches!(c, '.' | '?' | ';'))
+                    .map(|sentence| sentence.trim().to_string())
+                    .filter(|sentence| !sentence.is_empty())
+            })
+            .collect()
+    }
+    
+    fn clean_sentence(&self, sentence: &str) -> Vec<String> {
+        sentence
+            .chars()
+            .map(|c| {
+                // Keep 's as part of words, remove other punctuation
+                if c.is_alphabetic() || c.is_whitespace() || c == '\'' {
+                    c
+                } else {
+                    ' '
+                }
+            })
+            .collect::<String>()
+            .split_whitespace()
+            .map(|word| word.to_lowercase())
+            .filter(|word| !word.is_empty())
+            .collect()
+    }
+    
+    fn generate_substrings(&self, words: &[String]) -> Vec<Vec<String>> {
+        (0..words.len())
+            .flat_map(|start| {
+                ((start + 2)..=words.len())
+                    .map(move |end| words[start..end].to_vec())
+            })
+            .collect()
+    }
+    
+    fn clean_text(&self, text: &str) -> String {
+        text.chars()
+            .map(|c| {
+                // Keep 's as part of words, remove other punctuation
+                if c.is_alphabetic() || c.is_whitespace() || c == '\'' {
```

Comment:
c.is_alphabetic() || c.is_whitespace() || c == '\'' and the lines around it are repeated. Refactor to avoid that.